source/cbor.c

```diff
@@ -0,0 +1,739 @@
+        int error =                                                                                                    \
+            aws_byte_buf_ensure_capacity((encoder)->encoded_buf, (encoder)->encoded_buf->len + (length_to_reserve));   \
+        (void)error;                                                                                                   \
+        AWS_ASSERT(error == AWS_ERROR_SUCCESS);                                                                        \
+        size_t encoded_len = fn(value, s_get_encoder_current_position(encoder), s_get_encoder_remaining_len(encoder)); \
+        AWS_ASSERT((encoded_len) != 0);                                                                                \
+        (encoder)->encoded_buf->len += (encoded_len);                                                                  \
+    } while (false)
+
+void aws_cbor_encoder_write_uint(struct aws_cbor_encoder *encoder, uint64_t value) {
+    ENCODE_THROUGH_LIBCBOR(encoder, s_cbor_element_width_64bit, value, cbor_encode_uint);
+}
+
+void aws_cbor_encoder_write_negint(struct aws_cbor_encoder *encoder, uint64_t value) {
+    ENCODE_THROUGH_LIBCBOR(encoder, s_cbor_element_width_64bit, value, cbor_encode_negint);
+}
+
+void aws_cbor_encoder_write_single_float(struct aws_cbor_encoder *encoder, float value) {
+    ENCODE_THROUGH_LIBCBOR(encoder, s_cbor_element_width_32bit, value, cbor_encode_single);
+}
+
+void aws_cbor_encoder_write_float(struct aws_cbor_encoder *encoder, double value) {
+    if (!isfinite(value)) {
+        /* For special value: NAN/INFINITY, type cast to float and encode into single float. */
+        aws_cbor_encoder_write_single_float(encoder, (float)value);
+        return;
+    }
+    /* Conversation from int to floating-type is implementation defined if loss of precision */
+    if (value <= (double)INT64_MAX && value >= (double)INT64_MIN) {
+        /**
+         * A prvalue of a floating point type can be converted to a prvalue of an integer type. The conversion
+         * truncates; that is, the fractional part is discarded. The behavior is undefined if the truncated value cannot
+         * be represented in the destination type.
+         * Check against the INT64 range to avoid undefined behavior
+         *
+         * Comparing against INT64_MAX instead of UINT64_MAX to simplify the code, which may loss the opportunity to
+         * convert the UIN64 range from double to uint64_t. However, converting double to uint64_t will not benefit the
+         * total length encoded.
+         **/
+        int64_t int_value = (int64_t)value;
+        if (value == (double)int_value) {
+            if (int_value < 0) {
+                aws_cbor_encoder_write_negint(encoder, (uint64_t)(-1 - int_value));
+            } else {
+                aws_cbor_encoder_write_uint(encoder, (uint64_t)(int_value));
+            }
+            return;
+        }
+    }
+    if (value <= FLT_MAX && value >= -FLT_MAX) {
```

Comment:
It's because of undefined behavior. UBSAN will flag this when the conversion is outside the range. 

`FLT_MIN` is "Minimum representable **positive** floating-point number.". Only reason I put this check is to avoid UBSAN, I have tried [this](https://github.com/awslabs/aws-c-common/pull/1101/commits/f0b8a99ee0c4ead16df56c076d3314b464860048), but I am more happy about -max to max which is enough to silent the UBSAN.